### PR TITLE
Improvments to runtime configuration

### PR DESCRIPTION
**Pull request by karolk91**
## Summary

Add recursive wrapper call inspection to `ValidateStorageCalls` so that storage authorization is enforced even when `store`/`store_with_cid_config`/`renew` are nested inside `Utility::batch`, `Proxy::proxy`, `Sudo::sudo_as`, etc. Also harden XCM configuration to block storage-mutating calls over XCM, refactor `AllowedSignedCalls` into a single-source-of-truth allowlist, and add extensive test coverage for both runtimes.

### 1. `CallInspector` trait + `TraverseResult` (`pallets/transaction-storage/src/extension.rs`)

New trait that lets the runtime tell the pallet extension how to unwrap wrapper calls. The extension recursively walks the call tree and rejects any wrapper containing storage-mutating calls (store/renew must be direct extrinsics). For wrappers containing only management calls (authorize_*, refresh_*, remove_expired_*), it validates authorization and accumulates `ValidTransaction` metadata for correct mempool deduplication.

`TraverseResult` tracks:
- `found_storage` — whether any TransactionStorage pallet calls were found

### 2. Wrapper inspection helpers (`pallets/common/src/lib.rs`)

Generic helper functions for `pallet-utility`, `pallet-sudo`, and `pallet-proxy` that extract inner calls from wrapper call variants. Returns `None` for non-wrapper variants (e.g. `set_key`, `add_proxy`). Exhaustively matches all variants (including `__Ignore`) to ensure new upstream variants cause compile errors.

### 3. `StorageCallInspector` (both runtimes)

Runtime-specific `CallInspector` implementation. Also implements `Contains<RuntimeCall>` for use as XCM `SafeCallFilter`.

- **Polkadot**: inspects Utility, Proxy, and Sudo wrappers
- **Westend**: inspects Utility only. Sudo is intentionally not inspected so the sudo key holder can `sudo(store)` without authorization (Root origin is accepted by `ensure_authorized`)

### 4. `AllowedSignedCalls` refactor (Polkadot runtime)

Split into `validate_signed_call()` (recursive allowlist + auth checks) and a priority/longevity assignment block. Adds dedicated priority/longevity constants for Sudo, Proxy, Utility, and Upgrade calls (previously they all borrowed `BridgeTxLongevity`).

### 5. XCM `SafeCallFilter` (both runtimes)

Changed from `Everything` to `EverythingBut<StorageCallInspector>`, blocking `store`/`store_with_cid_config`/`renew` over XCM (including when wrapped in batch). Authorization management calls (`authorize_account`, etc.) remain allowed over XCM.

### Signed transactions — storage calls

| Scenario | Result | Reason |
|----------|--------|--------|
| Direct `store`/`renew` with authorization | **OK** | Origin transformed to `Authorized` |
| Direct `store`/`renew` without authorization | **Rejected** (`Payment`) | No auth found |
| `batch([store, ...])` with or without auth | **Rejected** (`Call`) | Store/renew must be direct extrinsics |
| `sudo(store)` (Westend only, sudo key holder) | **OK** | Sudo not inspected; Root accepted by `ensure_authorized` |
| `sudo(store)` / `sudo_as(store)` (Polkadot) | **Rejected** (`Call`) | Sudo inspected; store inside wrapper rejected |
| `proxy(store)` (Polkadot) | **Rejected** (`Call`) | Proxy inspected; store inside wrapper rejected |
| `batch_all([authorize_account])` by Authorizer | **OK** | Management calls allowed in wrappers |
| `batch([authorize_account])` by non-Authorizer | OK at validation, **fails at dispatch** | Dispatch checks Authorizer origin |
| Nesting > `MAX_WRAPPER_DEPTH` (8) | **Rejected** (`ExhaustsResources`) | Depth limit prevents complexity attacks |

### XCM Transact

| Scenario | Result |
|----------|--------|
| `store` / `store_with_cid_config` / `renew` | **Blocked** by `SafeCallFilter` |
| `batch([store])` | **Blocked** (recursive inspection) |
| `authorize_account` / `authorize_preimage` | **Allowed** (Superuser → Root satisfies Authorizer) |

### Polkadot solochain signed call allowlist

Allowed: `TransactionStorage::*`, `Session::set_keys/purge_keys`, bridge calls (`submit_finality_proof`, `submit_parachain_heads`, `receive_messages_proof/delivery_proof`), `Sudo::*`, `Proxy::*`, `Utility::*` (inner calls recursively validated), `System::apply_authorized_upgrade`.

Everything else (e.g., `System::remark`) is rejected with `InvalidTransaction::Call`, both direct and wrapped.

## Design: store/renew must be direct extrinsics

The `ValidateStorageCalls` extension rejects store/renew inside any wrapper. This is because:

- The extension transforms the origin to `Origin::Authorized { who, scope }` so that `store`/`renew` dispatches pass `ensure_authorized`
- Only wrappers that preserve the caller's origin (batch) could forward `Authorized`, but mixing storage + non-storage calls in a batch creates an origin conflict (`Authorized` vs `Signed`)
- Wrappers that change the origin (sudo, proxy) would lose the `Authorized` origin at dispatch
- Rejecting all wrapped store/renew is **fail-safe by design** — simpler and eliminates edge cases

**Exception**: On Westend, sudo is not inspected. `sudo(store)` works because sudo dispatches with Root origin, which `ensure_authorized` accepts directly (no `Authorized` origin needed).

## Test coverage

~1250 new lines of tests across both runtimes covering:
- Authorization enforcement for all wrapper types (batch, batch_all, force_batch, as_derivative, sudo_as, proxy)
- Store/renew rejection inside wrappers even when authorized
- Mixed batch rejection (storage + management, storage + non-storage)
- XCM SafeCallFilter for direct and wrapped storage calls
- XCM authorization management pass-through
- Recursion depth enforcement
- Call allowlist enforcement in wrappers (Polkadot)
- Authorization consumption verification (not consumed on rejection)
- `sudo(store)` works for sudo key holder on Westend

### CODE REVIEW

File: runtimes/bulletin-polkadot/tests/tests.rs [hunk 1]
```diff
@@ -1063,3 +1066,404 @@ fn allowed_signed_calls_preserves_storage_priority() {
...
+			TransactionValidityError::Invalid(InvalidTransaction::Call),
+			"System::remark",
+		);
+
+		// Note: sudo_as and proxy are NOT tested here because they have their own
+		// authorization mechanisms (sudo key, proxy delegation) and are intentionally
+		// permitted at the validation level. The pallet-side defense-in-depth checks
+		// protect against unauthorized operations through those paths.
+	});
+}
```
Comment (franciscoaguirre) [inline runtimes/bulletin-polkadot/tests/tests.rs]: I would add extra tests mixing store(account)/store(preimage)/renew/authorize in a single batch to see how that would work.

I would also add a test checking the maximum recursion depth

File: runtimes/bulletin-westend/src/xcm_config.rs [hunk 2]
```diff
@@ -226,6 +226,39 @@ pub type TrustedAliasers = (
 	AuthorizedAliasers<Runtime>,
 );
 
+/// Calls that are safe to dispatch from XCM. Blocks storage-mutating
+/// TransactionStorage calls — those require on-chain authorization that XCM cannot provide.
+/// Recursively inspects wrapper calls (Utility, Sudo) to prevent bypass via nesting.
+pub struct XcmSafeCallFilter;
```
[the remark this answers is not in the document]
Comment (karolk91) [reply runtimes/bulletin-westend/src/xcm_config.rs]: nice, added in https://github.com/paritytech/polkadot-bulletin-chain/pull/265/commits/6d414a92c4120354116606ac2e6d3cc4b252450f

File: pallets/transaction-storage/src/extension.rs [hunk 3]
```diff
@@ -64,7 +145,47 @@ impl<T: Config + Send + Sync> fmt::Debug for ValidateStorageCalls<T> {
...
+where
+	RuntimeCallOf<T>: IsSubType<Call<T>>,
+{
+	/// Recursively traverse a call tree, applying `visitor` to each storage call found.
+	///
+	/// Returns [`TraverseResult`] with:
+	/// - `found_storage`: whether any storage calls were visited
+	/// - `preserves_origin`: whether the outermost wrapper preserves the caller's origin
+	/// - `has_non_storage`: whether any non-storage, non-wrapper calls were found
+	fn traverse_storage_calls(
```
[the remark this answers is not in the document]
Comment (karolk91) [reply pallets/transaction-storage/src/extension.rs]: done
